### PR TITLE
add a flippable common fields validation before updaing a resource

### PR DESCRIPTION
**Pull request by caesarxuchao**
Ref #14126

This is the first step of doing validation for common fields. The validation turned off by default.

@jiangyaoguo @lavalamp @saad-ali @vishh @bgrant0607

### CODE REVIEW

**Comment by bgrant0607** [comment]
We plan to turn this on during testing?

What about PATCH?

File: cmd/kube-apiserver/app/server.go [hunk 1]
```diff
@@ -240,6 +241,8 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.KubeletConfig.CertFile, "kubelet-client-certificate", s.KubeletConfig.CertFile, "Path to a client cert file for TLS.")
 	fs.StringVar(&s.KubeletConfig.KeyFile, "kubelet-client-key", s.KubeletConfig.KeyFile, "Path to a client key file for TLS.")
 	fs.StringVar(&s.KubeletConfig.CAFile, "kubelet-certificate-authority", s.KubeletConfig.CAFile, "Path to a cert. file for the certificate authority.")
+	// TODO: delete this flag when we enable global validation by default. See #14126
+	fs.BoolVar(&rest.CommonUpdateValidation, "common-update-validation", false, "If true, validation of common fields, like UID, will be enabled for all update (PUT) requests.")
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: I would prefer to make this flag a "be nice to non-compliant clients" flag. We should always do the validation. The question is whether we reject bad requests or fix them.

File: pkg/api/rest/update.go [hunk 2]
```diff
@@ -44,6 +48,22 @@ type RESTUpdateStrategy interface {
...
+// TODO: add other common fields that require global validation.
+func commonFieldsValidation(obj, old runtime.Object) error {
+	objectMeta, err := api.ObjectMetaFor(obj)
+	if err != nil {
+		return errors.NewInternalError(err)
+	}
+	oldObjectMeta, err := api.ObjectMetaFor(old)
+	if err != nil {
+		return errors.NewInternalError(err)
+	}
```
Comment (lavalamp) [inline pkg/api/rest/update.go]: Maybe just call the validate object meta update function here instead of reimplementing.

**Comment by lavalamp** [comment]
The plan was to have a "be nice to bad clients" flag that we leave off in testing (so we find those clients) and just remove by 1.2, so people have time to fix their clients. (we'd want to announce this if we think such clients are common.)

PATCH should Just Work. And if a client changes a UID with a PATCH that should be rejected.

**Comment by caesarxuchao** [comment]
@lavalamp comments addressed. PTAL.

File: cmd/kube-apiserver/app/server.go [hunk 3]
```diff
@@ -240,6 +241,8 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.KubeletConfig.CertFile, "kubelet-client-certificate", s.KubeletConfig.CertFile, "Path to a client cert file for TLS.")
 	fs.StringVar(&s.KubeletConfig.KeyFile, "kubelet-client-key", s.KubeletConfig.KeyFile, "Path to a client key file for TLS.")
 	fs.StringVar(&s.KubeletConfig.CAFile, "kubelet-certificate-authority", s.KubeletConfig.CAFile, "Path to a cert. file for the certificate authority.")
+	// TODO: delete this flag when we stop fixing update requests at server-side. See #14126
+	fs.BoolVar(&validation.FixUpdateAtServer, "fix-update", true, "If true, server will do its best to fix the update request to pass the validation, e.g., setting empty UID in update request to its existing value.")
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: how about "repair-malformed-updates" as a name? and in the text, say we expect to delete the flag as soon as we identify and fix all such clients.

Default to false--let's see if the PR builder is happy. If is, maybe we don't need to flag protect at all.

**Comment by caesarxuchao** [comment]
> PATCH should Just Work. And if a client changes a UID with a PATCH that should be rejected.

It's correct. PATCH uses update under the hood and will be checked by the same validation functions.

**Comment by caesarxuchao** [comment]
Done. Let's see if it passes tests.

**Comment by caesarxuchao** [comment]
The integration test catches the empty UID of Event. So we need to use the flag to enable the fixing of malformed updates.

```
Server rejected event '&api.Event{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:api.ObjectMeta{Name:"node-scheduling-test-pod.140acca67e1e17cc", GenerateName:"", Namespace:"default", SelfLink:"", UID:"", ResourceVersion:"191", Generation:0, CreationTimestamp:unversioned.Time{Time:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*unversioned.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil)}, InvolvedObject:api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"node-scheduling-test-pod", UID:"c7f5d4e4-6caa-11e5-a6a4-0242ac1122d9", APIVersion:"v1", ResourceVersion:"190", FieldPath:""}, Reason:"FailedScheduling", Message:"no nodes available to schedule pods", Source:api.EventSource{Component:"scheduler", Host:""}, FirstTimestamp:unversioned.Time{Time:time.Time{sec:63579788445, nsec:0, loc:(*time.Location)(0x1f6ade0)}}, LastTimestamp:unversioned.Time{Time:time.Time{sec:63579788446, nsec:981742164, loc:(*time.Location)(0x1f6ade0)}}, Count:2}': 'Event "node-scheduling-test-pod.140acca67e1e17cc" is invalid: [uid: invalid value '', Details: field is immutable, uid: invalid value '', Details: field is immutable]' (will not retry!)
```

File: cmd/kube-apiserver/app/server.go [hunk 4]
```diff
@@ -245,6 +246,8 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.KubeletConfig.CAFile, "kubelet-certificate-authority", s.KubeletConfig.CAFile, "Path to a cert. file for the certificate authority.")
 	//See #14282 for details on how to test/try this option out.  TODO remove this comment once this option is tested in CI.
 	fs.IntVar(&s.KubernetesServiceNodePort, "kubernetes-service-node-port", 0, "If non-zero, the Kubernetes master service (which apiserver creates/maintains) will be of type NodePort, using this as the value of the port. If zero, the Kubernetes master service will be of type ClusterIP.")
+	// TODO: delete this flag as soon as we identify and fix all clients that send malformed updates, like #14126.
+	fs.BoolVar(&validation.FixUpdateAtServer, "repair-malformed-updates", false, "If true, server will do its best to fix the update request to pass the validation, e.g., setting empty UID in update request to its existing value.")
```
Comment (derekwaynecarr) [inline cmd/kube-apiserver/app/server.go]: should the flag be marked as deprecrated from the start?

Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: I guess we have to deal with this as long as there might be old kubelets.
We should note in the flag help that you need to have it on if you have
kubelets older than XXXX.

Comment (caesarxuchao) [inline cmd/kube-apiserver/app/server.go]: a nit-pick: not only old kubelets, also anything that uses old client. (e.g. the scheduler also generates Events so it  needs to be recompiled as well)

**Comment by caesarxuchao** [comment]
@lavalamp, I pushed another commit to removed  a bunch of calls to ValidateObjectMetaUpdate scattered in update validations for specific resources. PTAL. Thanks.

**Comment by caesarxuchao** [comment]
@lavalamp, I revert the last commit that removes scattered calls to ValidateObjectMetaUpdate, because that commit requires fixes to several unit tests. So I left a TODO and will cleaning it up in a following PR.

**Comment by caesarxuchao** [comment]
Ping @lavalamp

File: pkg/api/validation/validation.go [hunk 5]
```diff
@@ -38,6 +38,10 @@ import (
 	"github.com/golang/glog"
 )
 
+// TODO: delete this global variable when we enable the validation of common
+// fields by default.
+var FixUpdateAtServer bool = true
```
Comment (lavalamp) [inline pkg/api/validation/validation.go]: RepairMalformedUpdates

File: pkg/api/rest/update.go [hunk 6]
```diff
@@ -44,6 +45,22 @@ type RESTUpdateStrategy interface {
 	AllowUnconditionalUpdate() bool
 }
 
+// TODO: add other common fields that require global validation.
+func commonFieldsValidation(obj, old runtime.Object) fielderrors.ValidationErrorList {
```
Comment (lavalamp) [inline pkg/api/rest/update.go]: validateCommonFields

File: pkg/api/rest/update.go [hunk 7]
```diff
@@ -59,8 +76,13 @@ func BeforeUpdate(strategy RESTUpdateStrategy, ctx api.Context, obj, old runtime
 	} else {
 		objectMeta.Namespace = api.NamespaceNone
 	}
+
+	// Ensure some common fields, like UID, are validated for all resources.
+	errs := commonFieldsValidation(obj, old)
+
 	strategy.PrepareForUpdate(obj, old)
```
Comment (lavalamp) [inline pkg/api/rest/update.go]: Should this be called before the common field validation?

**Comment by lavalamp** [comment]
so the todo list after this goes in:
* cherry-pick to 1.1
* once your event PATCH change is merged, make another PR to turn this off in head.
* ... but leave it on in 1.1!

File: pkg/api/rest/update.go (hunk 7 above)
Comment (caesarxuchao) [inline pkg/api/rest/update.go]: Good catch. Thank you!

File: pkg/api/validation/validation.go (hunk 5 above)
Comment (caesarxuchao) [inline pkg/api/validation/validation.go]: Done. Sorry I was too lazy to fix this as I thought the variable will be short lived.

File: pkg/api/validation/validation.go [hunk 8]
```diff
@@ -273,8 +278,12 @@ func ValidateObjectMeta(meta *api.ObjectMeta, requiresNamespace bool, nameFn Val
 func ValidateObjectMetaUpdate(new, old *api.ObjectMeta) errs.ValidationErrorList {
 	allErrs := errs.ValidationErrorList{}
 
+	if !RepairMalformedUpdates && new.UID != old.UID {
+		allErrs = append(allErrs, errs.NewFieldInvalid("uid", new.UID, "field is immutable"))
+	}
 	// in the event it is left empty, set it, to allow clients more flexibility
-	if len(new.UID) == 0 {
+	// TODO: remove this logic when we retire the clients that modify UID.
```
Comment (lavalamp) [inline pkg/api/validation/validation.go]: sorry, one more request, add:
// Please do not copy this pattern elsewhere; validation functions should *not* be modifying the objects they are passed!

File: cmd/kube-apiserver/app/server.go [hunk 9]
```diff
@@ -245,6 +246,8 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.KubeletConfig.CAFile, "kubelet-certificate-authority", s.KubeletConfig.CAFile, "Path to a cert. file for the certificate authority.")
 	//See #14282 for details on how to test/try this option out.  TODO remove this comment once this option is tested in CI.
 	fs.IntVar(&s.KubernetesServiceNodePort, "kubernetes-service-node-port", 0, "If non-zero, the Kubernetes master service (which apiserver creates/maintains) will be of type NodePort, using this as the value of the port. If zero, the Kubernetes master service will be of type ClusterIP.")
+	// TODO: delete this flag as soon as we identify and fix all clients that send malformed updates, like #14126.
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: Please file an issue about removing this flag from head so we don't forget.

File: pkg/api/validation/validation.go [hunk 10]
```diff
@@ -273,22 +278,29 @@ func ValidateObjectMeta(meta *api.ObjectMeta, requiresNamespace bool, nameFn Val
...
-	if !old.DeletionTimestamp.IsZero() {
-		new.DeletionTimestamp = old.DeletionTimestamp
+	if !RepairMalformedUpdates && new.UID != old.UID {
+		allErrs = append(allErrs, errs.NewFieldInvalid("uid", new.UID, "field is immutable"))
 	}
-	if old.DeletionGracePeriodSeconds != nil && new.DeletionGracePeriodSeconds == nil {
-		new.DeletionGracePeriodSeconds = old.DeletionGracePeriodSeconds
+	// in the event it is left empty, set it, to allow clients more flexibility
+	// TODO: remove the following code that repairs the update request when we retire the clients that modify the immutable fields.
+	// Please do not copy this pattern elsewhere; validation functions should not be modifying the objects they are passed!
```
Comment (caesarxuchao) [inline pkg/api/validation/validation.go]: @lavalamp comments added. And I increase the scope of the `if` to enclose other code that repairs the update request.

Comment (lavalamp) [inline pkg/api/validation/validation.go]: Huh, I didn't realize deletion time stamp did the same thing.

File: pkg/api/validation/validation.go [hunk 11]
```diff
@@ -273,22 +278,29 @@ func ValidateObjectMeta(meta *api.ObjectMeta, requiresNamespace bool, nameFn Val
...
+		}
+		// an object can never remove a deletion timestamp or clear/change grace period seconds
+		if !old.DeletionTimestamp.IsZero() {
+			new.DeletionTimestamp = old.DeletionTimestamp
+		}
+		if old.DeletionGracePeriodSeconds != nil && new.DeletionGracePeriodSeconds == nil {
+			new.DeletionGracePeriodSeconds = old.DeletionGracePeriodSeconds
+		}
 	}
 	if new.DeletionGracePeriodSeconds != nil && old.DeletionGracePeriodSeconds != nil && *new.DeletionGracePeriodSeconds != *old.DeletionGracePeriodSeconds {
```
Comment (lavalamp) [inline pkg/api/validation/validation.go]: This check seems wrong. (what if new is nil but not old?) Can you file an issue?

**Comment by lavalamp** [comment]
LGTM

File: pkg/api/validation/validation.go (hunk 11 above)
Comment (caesarxuchao) [inline pkg/api/validation/validation.go]: that's handled by the repair in above. I'll leave a comment here so we don't forget when we delete the repair.

**Comment by caesarxuchao** [comment]
squashed and add a todo

File: pkg/api/validation/validation.go [hunk 12]
```diff
@@ -273,23 +278,32 @@ func ValidateObjectMeta(meta *api.ObjectMeta, requiresNamespace bool, nameFn Val
...
+		// an object can never remove a deletion timestamp or clear/change grace period seconds
+		if !old.DeletionTimestamp.IsZero() {
+			new.DeletionTimestamp = old.DeletionTimestamp
+		}
+		if old.DeletionGracePeriodSeconds != nil && new.DeletionGracePeriodSeconds == nil {
+			new.DeletionGracePeriodSeconds = old.DeletionGracePeriodSeconds
+		}
 	}
+
+	// TODO: needs to check if new==nil && old !=nil after the repair logic is removed.
```
Comment (caesarxuchao) [inline pkg/api/validation/validation.go]: added a todo here

**Comment by caesarxuchao** [comment]
add "lgtm" back since it's just squash

**Comment by gmarek** [comment]
@k8s-bot test this please

**Pull request merged**